### ISSUE TITLE
Zones/Ranks not being loaded in on first map after server restart/crash

### ISSUE THREAD

**Issue by Technoblazed** [jonitaikaponi/ckSurf#77]
The starting map that the server launches with after each restart and/or crash appears to not be able to connect to the database. The only way to resolve this issue is by admin restarting the map, or rtving onto the following map.

No errors are shown within the error logs.

**Comment by blackhawk74** [comment]
I have also run into the same issue. On server start, first map loads no zones/ranks. Every other map following is fine.

**Comment by Serg37SLT** [comment]
Also having the same issue here. Initial map, nothing loads. Further map changes everything is fine.

**Comment by gerrard9099** [comment]
Same issue here.

**Comment by ch4o5** [comment]
Same issue here as well. Was working previously, not sure which build introduced this...

**Comment by zAfLu** [comment]
can confirm that issue

**Comment by 4everGhost** [comment]
I can confirm this too.

Here is my small workaround, maybe it helps someone.

1. Your server needs to run in a tmux session, this is default for the "Server Management Script" by Daniel Gibb. I don't know about others.


2. Create a bash file next to the server script and insert the following code: 

cksurf_fix

```
#!/bin/bash
# Changes the map for the first tmux session

tmux send-keys "map surf_beginner" enter
```
This code needs to be run with the same user as your csgo server is running.


3. Open the crontab of the user the server is running for: for example `sudo -u server3 crontab -e`

```
# m h  dom mon dow   command
35 04 * * * /home/server3/servers/csgoservertest/csgoserver update-restart >/dev/null 2>&1
37 04 * * * /home/server3/servers/csgoservertest/cksurf_fix >/dev/null 2>&1

```

So after every restart my server is changing the map 3 minutes later. This workaround is quick and dirty and might not work for you but maybe it helps someone.

Cheers
Gh^2t

**Comment by Technoblazed** [comment]
@4everghosf the problem with this is that it's linux only. I run 64 slot servers and they just won't run smoothly on linux's single core per server. So I use windows instead.

**Comment by ch4o5** [comment]
Here's a small plugin which will restart the current map after 2 mins after server restart - should work on Windows aswell ;)

`#include <sourcemod>
#include <sdktools>
#define PLUGIN_VERSION "1.0.0"

public Plugin:myinfo =
{
	name = "ckSurf fix",
	author = "ch4os",
	description = "Reloads current map within 2 minutes after server restart",
	version = PLUGIN_VERSION,
	url = "www.killerspielplatz.com"
}

public OnPluginStart()
{
	CreateTimer(120.0, Event_ReloadMap);
}

public Action:Event_ReloadMap(Handle:Timer)
{
	decl String:MapName[255];
	GetCurrentMap(MapName, 255);
	PrintToServer("[ckSurf fix] Restarting current map...");
	ServerCommand("changelevel %s", MapName);
}`

**Comment by blackhawk74** [comment]
@ch4o5 
Nice temp fix! However, if someone joins an empty server, waking it from hibernation, it restarts the map, regardless of a fresh server restart/crash or not. Is there any way to make this not happen, and only on server start?

**Comment by marcowmadeira** [comment]
@blackhawk74 You can disable the hibernation.

`sv_hibernate_when_empty 0`

**Comment by blackhawk74** [comment]
@marcowmadeira

Right, but without disabling hibernation? Hibernation keeps my cpu loads down

**Comment by ch4o5** [comment]
Didn't think about hibernation, sorry. I don't use that feature. Timers don't work if hibernation is active. You could try this version instead (untested!):

```
#include <sourcemod>
#define PLUGIN_VERSION "1.0.1"

public Plugin:myinfo =
{
    name = "ckSurf fix",
    author = "ch4os",
    description = "Reloads current map within 2 minutes after server restart",
    version = PLUGIN_VERSION,
    url = "www.killerspielplatz.com"
}

new Handle:g_cvar_sv_hibernate_when_empty = INVALID_HANDLE;
new bool:g_useHibernate = false;

public OnPluginStart()
{
    g_cvar_sv_hibernate_when_empty = FindConVar("sv_hibernate_when_empty");

    if (GetConVarInt(g_cvar_sv_hibernate_when_empty) == 1) {
        SetConVarInt(g_cvar_sv_hibernate_when_empty, 0);
        g_useHibernate = true;
    }

    CreateTimer(120.0, Event_ReloadMap);
}

public Action:Event_ReloadMap(Handle:Timer)
{
    decl String:MapName[255];

    if (g_useHibernate)
        SetConVarInt(g_cvar_sv_hibernate_when_empty, 1);

    GetCurrentMap(MapName, 255);
    PrintToServer("[ckSurf fix] Restarting current map...");
    ServerCommand("changelevel %s", MapName);
}
```
This will check if sv_hibernate_when_empty is used and temporarily disable it until the timer activates. :)
